src/osd/PrimaryLogPG.cc

```diff
@@ -2607,11 +2608,51 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
     if (!chunk_data.length()) {
       return -ENODATA;
     }
-    tgt_length = chunk_data.length();
-    obj_op.add_data(CEPH_OSD_OP_WRITE, tgt_offset, tgt_length, chunk_data);
 
+    tgt_length = chunk_data.length();
     unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
-		     CEPH_OSD_FLAG_RWORDERED ;
+		     CEPH_OSD_FLAG_RWORDERED;
+    switch (manifest.fingerprint_mode) {
+      case object_manifest_t::TYPE_SHA1_FINGERPRINT:
+	{
+	  SHA1_Fingerprinter sha1_fp;
+	  object_t fp_oid = sha1_fp.do_fingerprint(chunk_data);
+	  cls_chunk_refcount_get_op call;
+	  bufferlist in;
+	  if (fp_oid != tgt_soid.oid) {
+	    // decrement old chunk's reference count 
+	    ObjectOperation dec_op;
+	    cls_chunk_refcount_put_op put_call;
+	    ::encode(put_call, in);                             
+	    dec_op.call("refcount", "chunk_put", in);         
+	    // we don't care dec_op's completion. scrub for dedup will fix this.
+	    tid = osd->objecter->mutate(
+	      tgt_soid.oid, oloc, dec_op, snapc,
+	      ceph::real_clock::from_ceph_timespec(obc->obs.oi.mtime),
+	      flags, NULL);
+	    in.clear();
+	  }
+	  tgt_soid.oid = fp_oid;
+	  iter->second.oid = tgt_soid;
+	  // add data op
+	  OSDOp& osd_op = obj_op.add_op(CEPH_OSD_OP_WRITE);
+	  osd_op.op.extent.offset = tgt_offset;
+	  osd_op.op.extent.length = tgt_length;
+	  osd_op.indata.claim_append(chunk_data);
+	  osd_op.op.flags = osd_op.op.flags | CEPH_OSD_OP_FLAG_WITH_DEDUP;
+	  // add refcount op
+	  call.source = tgt_soid;
```

Comment:
the source is *this* object, not the target object (named as the sha1 hash)

Reply:
Right.